eng/pipelines/templates/jobs/archetype-sdk-tests.yml

```diff
@@ -29,35 +29,92 @@ jobs:
+        Linux_x64_with_iot_samples:
+          OSVmImage: 'ubuntu-18.04'
+          vcpkg.deps: 'curl[ssl] cmocka paho-mqtt'
+          VCPKG_DEFAULT_TRIPLET: 'x64-linux'
+          build.args: ' -DTRANSPORT_PAHO=ON'
+          test_type: 'iot'
     pool:
       vmImage: $(OSVmImage)
 
     steps:
       - template: /eng/pipelines/templates/steps/vcpkg.yml
         parameters:
           DependenciesVariableName: vcpkg.deps
+          
+      - template: /eng/common/TestResources/deploy-test-resources.yml
+        parameters:
+          ServiceDirectory: '.'
+
+      - pwsh: |
+          # Generate certificate 
+          cd $SOURCES_DIR\sdk\samples\iot\
+          openssl ecparam -out device_ec_key.pem -name prime256v1 -genkey
+          openssl req -new -days 365 -nodes -x509 -key device_ec_key.pem -out device_ec_cert.pem -config x509_config.cfg -subj "/CN=$DEVICE_ID"
+
+          Get-Content device_ec_cert.pem, device_ec_key.pem | Set-Content device_cert_store.pem
+          openssl x509 -noout -fingerprint -in device_ec_cert.pem | % {$_.replace(":", "")} | % {$_.replace("SHA1 Fingerprint=", "")} | Tee-Object fingerprint.txt
+          $fingerprint = Get-Content -Path .\fingerprint.txt
+          
+          # Pass fingerprint to IoTHub 
+          Add-AzIotHubDevice `
+            -ResourceGroupName $ResourceGroupName `
+            -IotHubName $IOTHUB_NAME `
+            -DeviceId $DEVICE_ID `
+            -AuthMethod "x509_thumbprint" `
+            -PrimaryThumbprint $fingerprint`
+            -SecondaryThumbprint $fingerprint
+            
+          # Download Baltimore Cert
+          curl https://cacerts.digicert.com/BaltimoreCyberTrustRoot.crt.pem > $SOURCES_DIR\sdk\samples\iot\BaltimoreCyberTrustRoot.crt.pem
+          
+          # Link IoTHub to DPS service
+          hubConnectionString=$(az iot hub show-connection-string --name $IOTHUB_NAME --key primary --query connectionString -o tsv)
+          az iot dps linked-hub create --dps-name $DPS_NAME --resource-group $ResourceGroupName --connection-string $hubConnectionString --location eastus
+          
+        condition: contains(variables['test_type'], 'iot')
+        env:
+          VCPKG_DEFAULT_TRIPLET: x64-windows-static
+          VCPKG_ROOT: $(Agent.TempDirectory)
+          SOURCES_DIR: $(Build.SourcesDirectory)
+          DEVICE_ID: "aziotbld-c-sample"
+          IOTHUB_NAME: "aziotbld-embed-c"
+          DPS_NAME: "aziotbld-c-dps"
 
       - template: /eng/pipelines/templates/steps/cmake-build.yml
         parameters:
           BuildArgs: $(build.args)
 
-      - template: /eng/common/TestResources/deploy-test-resources.yml
-        parameters:
-          ServiceDirectory: '.'
-
       # Run live tests
       - ${{ each sample in parameters.LiveTestSamples }}:
         - script: ${{ sample.NonWindowsPath }}
           displayName: Live Test ${{ sample.Name }} (Non-Windows)
           condition: not(contains(variables['OSVmImage'], 'windows'))
-
+          env: 
```

Comment:
Do we need to pass these environment variables for all test matrix entries or only for iot type? Will these cause issues if they are set for other test runs? 

You can consider using setvariable (https://docs.microsoft.com/en-us/azure/devops/pipelines/process/variables?view=azure-devops&tabs=yaml%2Cbatch#set-a-job-scoped-variable-from-a-script) in your conditional step that sets these up so that they are only set in the env for the iot matrix entry.

Reply:
Just the iot tests, and only a subset of those, actually. Perhaps I could set a conditional about the tests by name?

Reply:
You cannot conditionalize the env variable section here. I'd suggest you set them at run time like the link I provided in my earlier comment.

Reply:
Yes they should be available and set as environment variables